src/RequestHandler/outofprocess/forwardinghandler.cpp

```diff
@@ -1492,76 +1575,94 @@ None
+    //
+    // Since we use TLS to guard WinHttp operation, call PostCompletion instead of
+    // IndicateCompletion to allow cleaning up the TLS before thread reuse.
+    // Never post after the request has been finished for whatever reason
+    //
+    // Only postCompletion after all WinHttp handle got closed,
+    // i.e., received WINHTTP_CALLBACK_STATUS_HANDLE_CLOSING callback for all handles
+    // So that no further WinHttp callback will be called
+    // Never post completion again after that
+    // Otherwise, there will be a AV as the request already passed IIS pipeline
+    //
+    if (fEndRequest && !m_fFinishRequest && dwHandlers == 0)
     {
-        DBG_ASSERT(TlsGetValue(g_dwTlsIndex) == this);
-        TlsSetValue(g_dwTlsIndex, NULL);
-        ReleaseSRWLockShared(&m_RequestLock);
-        DBG_ASSERT(TlsGetValue(g_dwTlsIndex) == NULL);
+        //
+        // Happy path
+        //
+        // Marked the request is finished, no more PostCompletion is allowed
+        RemoveRequest();
+        m_fFinishRequest = TRUE;
+        fDoPostCompletion = TRUE;
+        if (m_pWebSocket != NULL)
+        {
+            m_pWebSocket->Terminate();
+            m_pWebSocket = NULL;
+        }
     }
-
-    if (m_RequestStatus == FORWARDER_DONE)
+    else if (m_RequestStatus == FORWARDER_DONE)
     {
-        //disbale client disconnect callback
+        //
+        // Error path
+        //
         RemoveRequest();
-
-        if (m_hRequest != NULL)
+        if (m_hRequest != NULL && !m_fHttpHandleInClose)
         {
-            WinHttpSetStatusCallback(m_hRequest,
-                FORWARDING_HANDLER::OnWinHttpCompletion,
-                WINHTTP_CALLBACK_FLAG_HANDLES,
-                NULL);
-            if (WinHttpCloseHandle(m_hRequest))
-            {
-                m_hRequest = NULL;
-            }
-            else
-            {
-                // unexpected WinHttp error, log it
-                /*DebugBreak();
-                m_RequestStatus = FORWARDER_FINISH_REQUEST;
-                fDoPostCompletion = TRUE;*/
-            }
+            m_fHttpHandleInClose = TRUE;
+            WinHttpCloseHandle(m_hRequest);
+            m_hRequest = NULL;
         }
 
-        //
-        // If the request is a websocket request, initiate cleanup.
-        //
-        if (m_pWebSocket != NULL)
+        if (m_pWebSocket != NULL && !m_fWebSocketHandleInClose)
         {
+            m_fWebSocketHandleInClose = TRUE;
             m_pWebSocket->TerminateRequest();
         }
 
         if (fEndRequest)
         {
-            // only postCompletion after WINHTTP_CALLBACK_STATUS_HANDLE_CLOSING
-            // so that no further WinHttp callback will be called
-            // in case of websocket, m_hRequest has already been closed after upgrade
-            // websocket will handle completion
-            m_fFinishRequest = TRUE;
-            fDoPostCompletion = TRUE;
+            fDoPostCompletion = (dwHandlers == 0 && !m_fFinishRequest);
```

Comment:
nit: you don't need to check m_fFinishRequest here; you can always just set m_fFinishRequest to true if dwHandlers == 0.

Reply:
fixed